src/coprocessor/cache.rs

```diff
@@ -0,0 +1,493 @@
+// Copyright 2016 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+extern crate linked_hash_map;
+
+use std::collections::HashMap;
+use std::option::Option;
+use std::boxed::Box;
+use std::sync::{Arc, Mutex};
+use self::linked_hash_map::LinkedHashMap;
+use super::metrics::*;
+
+type DistSQLCacheKey = String;
+
+const DISTSQL_CACHE_ENTRY_ADDITION_SIZE: usize = 40;
+
+pub struct DistSQLCacheEntry {
+    region_id: u64,
+    version: u64,
+    result: Vec<u8>,
+    size: usize,
+}
+
+impl DistSQLCacheEntry {
+    pub fn new(region_id: u64, version: u64, key_size: usize, r: Vec<u8>) -> DistSQLCacheEntry {
+        let size = r.len() + DISTSQL_CACHE_ENTRY_ADDITION_SIZE + 2 * key_size;
+        DistSQLCacheEntry {
+            region_id: region_id,
+            version: version,
+            result: r,
+            size: size,
+        }
+    }
+}
+
+pub struct RegionDistSQLCacheEntry {
+    version: u64,
+    enable: bool,
+    cached_items: HashMap<DistSQLCacheKey, u8>,
+}
+
+pub struct DistSQLCache {
+    regions: HashMap<u64, RegionDistSQLCacheEntry>,
```

Comment:
It's not a good idea to split `regions` with the following `map`?  Should we remove related keys from `map` when the related region been removed?

Reply:
Split `regions` into another hash map is an optimization. If we got a region should be flush, we can use `regions` hash map to get the exact list instead iterate all cache entries.